syntax = "proto3";

import "data/common.proto";

package data;

option cc_enable_arenas = true;
option java_outer_classname = "FileSummaryProtos";

message FileSummary {
    enum FileOp {
        UNSET  = 0;
        CREATE = 1;
        READ   = 2;
        WRITE  = 3;
        OPEN   = 4;
        CLOSE  = 5;
        MOVE   = 6;
        REMOVE = 7;
        LINK   = 8;
        PERM   = 9;
        CWD    = 10;
    }

    message FileOperation {
        reserved 1; // Formerly used for syscall name
        FileOp operation = 2;
    }

    message Container {
        string id = 1;
        bool privileged = 2;
        UUID cluster_id = 3;
    }

    message Permissions {
        uint32 owner = 1;
        uint32 fs_group = 2;
        uint32 all_users = 3;
    }

    message File {
        string path = 1;
        Permissions permissions = 2;
        oneof set_owner_uid { // oneof since zero value has special meaning
            uint64 owner_uid = 3;
        }
        oneof set_owner_gid { // oneof since zero value has special meaning
            uint64 owner_gid = 4;
        }
        reserved 5; // Previously used for sensitive_contents.
        bool directory = 6;
        string new_path = 7;
    }

    message Process {
        string cmdline = 1;
        oneof set_uid { // oneof since zero value has special meaning
            uint64 uid = 2;
        }
        oneof set_gid { // oneof since zero value has special meaning
            uint64 gid = 3;
        }
    }

    UUID id = 1;
    Container container = 2;
    File file = 3;
    Process process = 4;
    repeated FileOperation ops = 5;
    int64 start_time = 6;
    int64 end_time = 7;
    Status status = 8;
}
